// @react-spectrum/well/dist/module.js
import _react, { forwardRef } from "react";
import { filterDOMProps } from "@react-aria/utils";
import { classNames, useDOMRef, useStyleProps } from "@react-spectrum/utils";
import _babelRuntimeHelpersEsmObjectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import _babelRuntimeHelpersEsmExtends from "@babel/runtime/helpers/esm/extends";
import "./main.css";

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

// ASSET: /Users/govett/dev/react-spectrum-v3/node_modules/@adobe/spectrum-css-temp/components/well/vars.css
var $da02da2493443b16432bdd47cac91318$exports = {};
$da02da2493443b16432bdd47cac91318$exports = {
  "spectrum-Well": "_spectrum-Well_daead"
};
var $da02da2493443b16432bdd47cac91318$$interop$default = $parcel$interopDefault($da02da2493443b16432bdd47cac91318$exports);

function $bd3e6f0b2b67e7ca0478475f0cdb9$var$Well(props, ref) {
  let {
    children,
    role
  } = props,
      otherProps = _babelRuntimeHelpersEsmObjectWithoutPropertiesLoose(props, ["children", "role"]);

  let domRef = useDOMRef(ref);
  let {
    styleProps
  } = useStyleProps(otherProps);

  if (!role && (props['aria-label'] || props['aria-labelledby'])) {
    console.warn('A labelled Well must have a role.');
  }

  return /*#__PURE__*/_react.createElement("div", _babelRuntimeHelpersEsmExtends({}, filterDOMProps(otherProps, {
    labelable: !!role
  }), styleProps, {
    role: role,
    ref: domRef,
    className: classNames($da02da2493443b16432bdd47cac91318$$interop$default, 'spectrum-Well', styleProps.className)
  }), children);
}
/**
 * A Well is a content container that displays non-editable content separate from other content on the screen.
 * Often this is used to display preformatted text, such as code/markup examples on a documentation page.
 */


export const Well = forwardRef($bd3e6f0b2b67e7ca0478475f0cdb9$var$Well);
