// @react-spectrum/utils/dist/main.js
var {
  mergeProps
} = require("@react-aria/utils");

var {
  useLocale
} = require("@react-aria/i18n");

var _babelRuntimeHelpersObjectWithoutPropertiesLoose = $parcel$interopDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _babelRuntimeHelpersExtends = $parcel$interopDefault(require("@babel/runtime/helpers/extends"));

var _react2 = require("react");

var _react = $parcel$interopDefault(_react2);

var {
  useEffect,
  useState,
  useImperativeHandle,
  useRef,
  useContext,
  useMemo,
  useLayoutEffect
} = _react2;

var _classnames = $parcel$interopDefault(require("classnames"));

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

let shouldKeepSpectrumClassNames = false;
exports.shouldKeepSpectrumClassNames = shouldKeepSpectrumClassNames;

function keepSpectrumClassNames() {
  shouldKeepSpectrumClassNames = true;
  exports.shouldKeepSpectrumClassNames = shouldKeepSpectrumClassNames;
  console.warn('Legacy spectrum-prefixed class names enabled for backward compatibility. ' + 'We recommend replacing instances of CSS overrides targeting spectrum selectors ' + 'in your app with custom class names of your own, and disabling this flag.');
}

exports.keepSpectrumClassNames = keepSpectrumClassNames;

function classNames(cssModule) {
  let classes = [];

  for (var _len = arguments.length, values = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    values[_key - 1] = arguments[_key];
  }

  for (let value of values) {
    if (typeof value === 'object' && value) {
      let mapped = {};

      for (let key in value) {
        if (cssModule[key]) {
          mapped[cssModule[key]] = value[key];
        }

        if (shouldKeepSpectrumClassNames || !cssModule[key]) {
          mapped[key] = value[key];
        }
      }

      classes.push(mapped);
    } else if (typeof value === 'string') {
      if (cssModule[value]) {
        classes.push(cssModule[value]);
      }

      if (shouldKeepSpectrumClassNames || !cssModule[value]) {
        classes.push(value);
      }
    } else {
      classes.push(value);
    }
  }

  return _classnames(...classes);
}

exports.classNames = classNames;

function getWrappedElement(children) {
  let element;

  if (typeof children === 'string') {
    element = /*#__PURE__*/_react.createElement("span", null, children);
  } else {
    element = _react.Children.only(children);
  }

  return element;
}

exports.getWrappedElement = getWrappedElement;

function useMediaQuery(query) {
  let supportsMatchMedia = typeof window !== 'undefined' && typeof window.matchMedia === 'function';
  let [matches, setMatches] = useState(() => supportsMatchMedia ? window.matchMedia(query).matches : false);
  useEffect(() => {
    if (!supportsMatchMedia) {
      return;
    }

    let mq = window.matchMedia(query);

    let onChange = evt => {
      setMatches(evt.matches);
    };

    mq.addListener(onChange);
    return () => {
      mq.removeListener(onChange);
    };
  }, [supportsMatchMedia, query]);
  return matches;
}

exports.useMediaQuery = useMediaQuery;

function createDOMRef(ref) {
  return {
    UNSAFE_getDOMNode() {
      return ref.current;
    }

  };
}

exports.createDOMRef = createDOMRef;

function createFocusableRef(domRef, focusableRef) {
  if (focusableRef === void 0) {
    focusableRef = domRef;
  }

  return _babelRuntimeHelpersExtends({}, createDOMRef(domRef), {
    focus() {
      if (focusableRef.current) {
        focusableRef.current.focus();
      }
    }

  });
}

exports.createFocusableRef = createFocusableRef;

function useDOMRef(ref) {
  let domRef = useRef(null);
  useImperativeHandle(ref, () => createDOMRef(domRef));
  return domRef;
}

exports.useDOMRef = useDOMRef;

function useFocusableRef(ref, focusableRef) {
  let domRef = useRef(null);
  useImperativeHandle(ref, () => createFocusableRef(domRef, focusableRef));
  return domRef;
}

exports.useFocusableRef = useFocusableRef;

function unwrapDOMRef(ref) {
  return {
    get current() {
      return ref.current && ref.current.UNSAFE_getDOMNode();
    }

  };
}

exports.unwrapDOMRef = unwrapDOMRef;
const baseStyleProps = {
  margin: ['margin', dimensionValue],
  marginStart: [$fde11e9d9d7d059795140092a2808e57$var$rtl('marginLeft', 'marginRight'), dimensionValue],
  marginEnd: [$fde11e9d9d7d059795140092a2808e57$var$rtl('marginRight', 'marginLeft'), dimensionValue],
  // marginLeft: ['marginLeft', dimensionValue],
  // marginRight: ['marginRight', dimensionValue],
  marginTop: ['marginTop', dimensionValue],
  marginBottom: ['marginBottom', dimensionValue],
  marginX: [['marginLeft', 'marginRight'], dimensionValue],
  marginY: [['marginTop', 'marginBottom'], dimensionValue],
  width: ['width', dimensionValue],
  height: ['height', dimensionValue],
  minWidth: ['minWidth', dimensionValue],
  minHeight: ['minHeight', dimensionValue],
  maxWidth: ['maxWidth', dimensionValue],
  maxHeight: ['maxHeight', dimensionValue],
  isHidden: ['display', $fde11e9d9d7d059795140092a2808e57$var$hiddenValue],
  alignSelf: ['alignSelf', passthroughStyle],
  justifySelf: ['justifySelf', passthroughStyle],
  position: ['position', $fde11e9d9d7d059795140092a2808e57$var$anyValue],
  zIndex: ['zIndex', $fde11e9d9d7d059795140092a2808e57$var$anyValue],
  top: ['top', dimensionValue],
  bottom: ['bottom', dimensionValue],
  start: [$fde11e9d9d7d059795140092a2808e57$var$rtl('left', 'right'), dimensionValue],
  end: [$fde11e9d9d7d059795140092a2808e57$var$rtl('right', 'left'), dimensionValue],
  left: ['left', dimensionValue],
  right: ['right', dimensionValue],
  flex: ['flex', $fde11e9d9d7d059795140092a2808e57$var$flexValue],
  flexGrow: ['flexGrow', passthroughStyle],
  flexShrink: ['flexShrink', passthroughStyle],
  flexBasis: ['flexBasis', passthroughStyle],
  gridArea: ['gridArea', passthroughStyle],
  gridColumn: ['gridColumn', passthroughStyle],
  gridColumnEnd: ['gridColumnEnd', passthroughStyle],
  gridColumnStart: ['gridColumnStart', passthroughStyle],
  gridRow: ['gridRow', passthroughStyle],
  gridRowEnd: ['gridRowEnd', passthroughStyle],
  gridRowStart: ['gridRowStart', passthroughStyle]
};
exports.baseStyleProps = baseStyleProps;

const viewStyleProps = _babelRuntimeHelpersExtends({}, baseStyleProps, {
  backgroundColor: ['backgroundColor', $fde11e9d9d7d059795140092a2808e57$var$backgroundColorValue],
  borderWidth: ['borderWidth', $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderStartWidth: [$fde11e9d9d7d059795140092a2808e57$var$rtl('borderLeftWidth', 'borderRightWidth'), $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderEndWidth: [$fde11e9d9d7d059795140092a2808e57$var$rtl('borderRightWidth', 'borderLeftWidth'), $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderLeftWidth: ['borderLeftWidth', $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderRightWidth: ['borderRightWidth', $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderTopWidth: ['borderTopWidth', $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderBottomWidth: ['borderBottomWidth', $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderXWidth: [['borderLeftWidth', 'borderRightWidth'], $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderYWidth: [['borderTopWidth', 'borderBottomWidth'], $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue],
  borderColor: ['borderColor', $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderStartColor: [$fde11e9d9d7d059795140092a2808e57$var$rtl('borderLeftColor', 'borderRightColor'), $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderEndColor: [$fde11e9d9d7d059795140092a2808e57$var$rtl('borderRightColor', 'borderLeftColor'), $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderLeftColor: ['borderLeftColor', $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderRightColor: ['borderRightColor', $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderTopColor: ['borderTopColor', $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderBottomColor: ['borderBottomColor', $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderXColor: [['borderLeftColor', 'borderRightColor'], $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderYColor: [['borderTopColor', 'borderBottomColor'], $fde11e9d9d7d059795140092a2808e57$var$borderColorValue],
  borderRadius: ['borderRadius', $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  borderTopStartRadius: [$fde11e9d9d7d059795140092a2808e57$var$rtl('borderTopLeftRadius', 'borderTopRightRadius'), $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  borderTopEndRadius: [$fde11e9d9d7d059795140092a2808e57$var$rtl('borderTopRightRadius', 'borderTopLeftRadius'), $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  borderBottomStartRadius: [$fde11e9d9d7d059795140092a2808e57$var$rtl('borderBottomLeftRadius', 'borderBottomRightRadius'), $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  borderBottomEndRadius: [$fde11e9d9d7d059795140092a2808e57$var$rtl('borderBottomRightRadius', 'borderBottomLeftRadius'), $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  borderTopLeftRadius: ['borderTopLeftRadius', $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  borderTopRightRadius: ['borderTopRightRadius', $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  borderBottomLeftRadius: ['borderBottomLeftRadius', $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  borderBottomRightRadius: ['borderBottomRightRadius', $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue],
  padding: ['padding', dimensionValue],
  paddingStart: [$fde11e9d9d7d059795140092a2808e57$var$rtl('paddingLeft', 'paddingRight'), dimensionValue],
  paddingEnd: [$fde11e9d9d7d059795140092a2808e57$var$rtl('paddingRight', 'paddingLeft'), dimensionValue],
  paddingLeft: ['paddingLeft', dimensionValue],
  paddingRight: ['paddingRight', dimensionValue],
  paddingTop: ['paddingTop', dimensionValue],
  paddingBottom: ['paddingBottom', dimensionValue],
  paddingX: [['paddingLeft', 'paddingRight'], dimensionValue],
  paddingY: [['paddingTop', 'paddingBottom'], dimensionValue],
  overflow: ['overflow', passthroughStyle]
});

exports.viewStyleProps = viewStyleProps;
const $fde11e9d9d7d059795140092a2808e57$var$borderStyleProps = {
  borderWidth: 'borderStyle',
  borderLeftWidth: 'borderLeftStyle',
  borderRightWidth: 'borderRightStyle',
  borderTopWidth: 'borderTopStyle',
  borderBottomWidth: 'borderBottomStyle'
};

function $fde11e9d9d7d059795140092a2808e57$var$rtl(ltr, rtl) {
  return direction => direction === 'rtl' ? rtl : ltr;
}

const $fde11e9d9d7d059795140092a2808e57$var$UNIT_RE = /(%|px|em|rem|vw|vh|auto|cm|mm|in|pt|pc|ex|ch|rem|vmin|vmax|fr)$/;

function dimensionValue(value) {
  if (typeof value === 'number') {
    return value + 'px';
  }

  if ($fde11e9d9d7d059795140092a2808e57$var$UNIT_RE.test(value)) {
    return value;
  }

  return "var(--spectrum-global-dimension-" + value + ", var(--spectrum-alias-" + value + "))";
}

exports.dimensionValue = dimensionValue;

function $fde11e9d9d7d059795140092a2808e57$var$colorValue(value, type) {
  if (type === void 0) {
    type = 'default';
  }

  return "var(--spectrum-global-color-" + value + ", var(--spectrum-semantic-" + value + "-color-" + type + "))";
}

function $fde11e9d9d7d059795140092a2808e57$var$backgroundColorValue(value) {
  return "var(--spectrum-alias-background-color-" + value + ", " + $fde11e9d9d7d059795140092a2808e57$var$colorValue(value, 'background') + ")";
}

function $fde11e9d9d7d059795140092a2808e57$var$borderColorValue(value) {
  if (value === 'default') {
    return 'var(--spectrum-alias-border-color)';
  }

  return "var(--spectrum-alias-border-color-" + value + ", " + $fde11e9d9d7d059795140092a2808e57$var$colorValue(value, 'border') + ")";
}

function $fde11e9d9d7d059795140092a2808e57$var$borderSizeValue(value) {
  return "var(--spectrum-alias-border-size-" + value + ")";
}

function $fde11e9d9d7d059795140092a2808e57$var$borderRadiusValue(value) {
  return "var(--spectrum-alias-border-radius-" + value + ")";
}

function $fde11e9d9d7d059795140092a2808e57$var$hiddenValue(value) {
  return value ? 'none' : undefined;
}

function $fde11e9d9d7d059795140092a2808e57$var$anyValue(value) {
  return value;
}

function $fde11e9d9d7d059795140092a2808e57$var$flexValue(value) {
  if (typeof value === 'boolean') {
    return value ? '1' : undefined;
  }

  return '' + value;
}

function convertStyleProps(props, handlers, direction) {
  let style = {};

  for (let key in props) {
    let styleProp = handlers[key];

    if (!styleProp || props[key] == null) {
      continue;
    }

    let [name, convert] = styleProp;

    if (typeof name === 'function') {
      name = name(direction);
    }

    let value = convert(props[key]);

    if (Array.isArray(name)) {
      for (let k of name) {
        style[k] = value;
      }
    } else {
      style[name] = value;
    }
  }

  for (let prop in $fde11e9d9d7d059795140092a2808e57$var$borderStyleProps) {
    if (style[prop]) {
      style[$fde11e9d9d7d059795140092a2808e57$var$borderStyleProps[prop]] = 'solid';
      style.boxSizing = 'border-box';
    }
  }

  return style;
}

exports.convertStyleProps = convertStyleProps;

function useStyleProps(props, handlers) {
  if (handlers === void 0) {
    handlers = baseStyleProps;
  }

  let {
    UNSAFE_className,
    UNSAFE_style
  } = props,
      otherProps = _babelRuntimeHelpersObjectWithoutPropertiesLoose(props, ["UNSAFE_className", "UNSAFE_style"]);

  let {
    direction
  } = useLocale();
  let styles = convertStyleProps(props, handlers, direction);

  let style = _babelRuntimeHelpersExtends({}, UNSAFE_style, styles); // @ts-ignore


  if (otherProps.className) {
    console.warn('The className prop is unsafe and is unsupported in React Spectrum v3. ' + 'Please use style props with Spectrum variables, or UNSAFE_className if you absolutely must to something custom. ' + 'Note that this may break in future versions due to DOM structure changes.');
  } // @ts-ignore


  if (otherProps.style) {
    console.warn('The style prop is unsafe and is unsupported in React Spectrum v3. ' + 'Please use style props with Spectrum variables, or UNSAFE_style if you absolutely must to something custom. ' + 'Note that this may break in future versions due to DOM structure changes.');
  }

  let styleProps = {
    style,
    className: UNSAFE_className
  };

  if (props.isHidden) {
    styleProps.hidden = true;
  }

  return {
    styleProps
  };
}

exports.useStyleProps = useStyleProps;

function passthroughStyle(value) {
  return value;
}

exports.passthroughStyle = passthroughStyle;

let $fda26b2aacd0a7728829de04d10c975$var$SlotContext = _react.createContext(null);

function useSlotProps(props, defaultSlot) {
  let slot = props.slot || defaultSlot;
  let {
    [slot]: slotProps = {}
  } = useContext($fda26b2aacd0a7728829de04d10c975$var$SlotContext) || {};
  return mergeProps(slotProps, props);
}

exports.useSlotProps = useSlotProps;

function cssModuleToSlots(cssModule) {
  return Object.keys(cssModule).reduce((acc, slot) => {
    acc[slot] = {
      UNSAFE_className: cssModule[slot]
    };
    return acc;
  }, {});
}

exports.cssModuleToSlots = cssModuleToSlots;

function SlotProvider(props) {
  let parentSlots = useContext($fda26b2aacd0a7728829de04d10c975$var$SlotContext) || {};
  let {
    slots = {},
    children
  } = props; // Merge props for each slot from parent context and props

  let value = useMemo(() => Object.keys(parentSlots).concat(Object.keys(slots)).reduce((o, p) => _babelRuntimeHelpersExtends({}, o, {
    [p]: mergeProps(parentSlots[p] || {}, slots[p] || {})
  }), {}), [parentSlots, slots]);
  return /*#__PURE__*/_react.createElement($fda26b2aacd0a7728829de04d10c975$var$SlotContext.Provider, {
    value: value
  }, children);
}

exports.SlotProvider = SlotProvider;

function ClearSlots(props) {
  let {
    children
  } = props,
      otherProps = _babelRuntimeHelpersObjectWithoutPropertiesLoose(props, ["children"]);

  let content = children;

  if (_react.Children.toArray(children).length <= 1) {
    if (typeof children === 'function') {
      // need to know if the node is a string or something else that react can render that doesn't get props
      content = _react.cloneElement(_react.Children.only(children), otherProps);
    }
  }

  return /*#__PURE__*/_react.createElement($fda26b2aacd0a7728829de04d10c975$var$SlotContext.Provider, {
    value: {}
  }, content);
}

exports.ClearSlots = ClearSlots;

function useHasChild(query, ref) {
  let [hasChild, setHasChild] = useState(true);
  useLayoutEffect(() => {
    setHasChild(!!(ref.current && ref.current.querySelector(query)));
  }, [setHasChild, query, ref]);
  return hasChild;
}

exports.useHasChild = useHasChild;
